### PR TITLE
Fix SN not backprop via sigma(W), and not reusing W_u

### PR DESCRIPTION
**Pull request by ssnl**
cc @crcrpar @t-vi

### CODE REVIEW

**Comment by crcrpar** [comment]
Seems to I don't understand gradient recording mechanism.

thank you very much.

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -40,8 +43,7 @@ def remove(self, module):
     def __call__(self, module, inputs):
         weight, u = self.compute_weight(module)
         setattr(module, self.name, weight)
-        with torch.no_grad():
-            getattr(module, self.name).copy_(weight)
+        setattr(module, self.name + '_u', u)
```
Comment (t-vi) [inline torch/nn/utils/spectral_norm.py]: Uh, yes, this needs to be `_u`, not `weight`. I seem to recall that there was a reason to use `copy_`, though. My impression is that this might behave differently across Python versions, in particular Python 3.6 vs. earlier. Is it OK to overwrite a registered buffer with a calculated variable?

File: torch/nn/utils/spectral_norm.py [hunk 2]
```diff
@@ -26,7 +29,7 @@ def compute_weight(self, module):
                 v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
                 u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
 
-            sigma = torch.dot(u, torch.matmul(weight_mat, v))
+        sigma = torch.dot(u, torch.matmul(weight_mat, v))
```
Comment (t-vi) [inline torch/nn/utils/spectral_norm.py]: Yes, please! We indeed missed the second term in the gradient before.

**Review by t-vi — COMMENTED** [review]
Thanks!

File: torch/nn/utils/spectral_norm.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: It should be okay. The `__setattr__` removes the current assigned values first anyways. According to local test with Py 3.6, it seems not leaking.

Comment (t-vi) [reply torch/nn/utils/spectral_norm.py]: Right. Thanks for checking. (What I was concerned about was whether there is an analogous issue to [the "cannot overwrite parameter with variable check"](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L549) only working in Python 3.6. But it's not the case here.

**Review by soumith — APPROVED** [review]
